src/object/cli_obj.c

```diff
@@ -2444,19 +2437,419 @@ obj_req_fanout(struct dc_object *obj, struct obj_auxi_args *obj_auxi,
+					       obj_ec_stripe_rec_nr(oca),
+					       obj_ec_cell_rec_nr(oca),
+					       shard_auxi->shard);
+
+		data_size = min(obj_ec_cell_rec_nr(oca), total_size);
+		total_size -= data_size;
+		D_DEBUG(DB_IO, "total "DF_U64" merge "DF_U64"/"DF_U64"\n",
+			total_size, daos_off, data_size);
+		rc = merge_recx(merge_list, daos_off, data_size);
+		if (rc)
+			break;
+
+		cur_off += data_size;
+	}
+	return rc;
+}
+
+static int
+obj_shard_list_recx_cb(struct shard_auxi_args *shard_auxi,
+		       struct obj_auxi_args *obj_auxi, void *arg)
+{
+	struct comp_cb_merge_arg	*merge_arg = arg;
+	struct shard_list_args		*shard_arg;
+	int i;
+
+	shard_arg = container_of(shard_auxi, struct shard_list_args, la_auxi);
+	/* convert recxs for EC object*/
+	for (i = 0; i < shard_arg->la_nr; i++) {
+		int rc;
+
+		rc = obj_ec_recxs_convert(&merge_arg->merge_list,
+					  &shard_arg->la_recxs[i],
+					  shard_auxi);
+		if (rc) {
+			if (obj_auxi->obj_task->dt_result == 0)
+				obj_auxi->obj_task->dt_result = rc;
+			D_ERROR(DF_OID" recx convert failed: %d\n",
+				DP_OID(obj_auxi->obj->cob_md.omd_id), rc);
+			return rc;
+		}
+	}
+
+	return 0;
+}
+
+static int
+obj_enum_shard_process_cb(daos_key_desc_t *kds, void *ptr,
+			  unsigned int size, void *arg)
+{
+	struct obj_enum_rec	*rec = ptr;
+	struct shard_auxi_args	*shard_auxi = arg;
+	int			rc;
+
+	D_ASSERT(size >= sizeof(struct obj_enum_rec));
+
+	rc = obj_ec_recxs_convert(NULL, &rec->rec_recx, shard_auxi);
+	if (rc)
+		return rc;
+
+	return rc;
+}
+
+static int
+obj_shard_list_obj_cb(struct shard_auxi_args *shard_auxi,
+		      struct obj_auxi_args *obj_auxi, void *arg)
+{
+	struct comp_cb_merge_arg *merge_arg = arg;
+	struct shard_list_args	*shard_arg;
+	daos_obj_list_t		*obj_arg = dc_task_get_args(obj_auxi->obj_task);
+	char			*ptr = obj_arg->sgl->sg_iovs[0].iov_buf;
+	daos_key_desc_t		*kds = obj_arg->kds;
+	int			rc;
+
+	shard_arg = container_of(shard_auxi, struct shard_list_args, la_auxi);
+	rc = obj_enum_iterate(shard_arg->la_kds, shard_arg->la_sgl,
+			      shard_arg->la_nr, OBJ_ITER_RECX,
+			      obj_enum_shard_process_cb, shard_auxi);
+	ptr += merge_arg->merge_sgl_off;
+	if (ptr != shard_arg->la_sgl->sg_iovs[0].iov_buf)
+		memmove(ptr, shard_arg->la_sgl->sg_iovs[0].iov_buf,
+			shard_arg->la_sgl->sg_iovs[0].iov_len);
+	obj_arg->sgl->sg_iovs[0].iov_len +=
+			shard_arg->la_sgl->sg_iovs[0].iov_len;
+	merge_arg->merge_sgl_off += shard_arg->la_sgl->sg_iovs[0].iov_len;
+
+	kds += merge_arg->merge_nr;
+	if (kds != shard_arg->la_kds) {
+		int i;
+
+		for (i = 0; i < shard_arg->la_nr; i++)
+			kds[i] = shard_arg->la_kds[i];
+	}
+	merge_arg->merge_nr += shard_arg->la_nr;
+
+	D_DEBUG(DB_TRACE, "merge_nr %d\n", merge_arg->merge_nr);
+	return rc;
+}
+
+static int
+merge_key(d_list_t *head, char *key, int key_size)
+{
+	struct obj_auxi_list_key *key_one;
+
+	d_list_for_each_entry(key_one, head, key_list) {
+		if (key_size == key_one->key.iov_len &&
+		    strncmp(key_one->key.iov_buf, key, key_size)) {
```

Comment:
&& strncmp(key_one->key.iov_buf, key, key_size) == 0 ?
BTW, seems it is inefficient as it need to compare all existing keys every time, maybe can use key hash? (can refine it in following patch)

Reply:
This is only needed for degraded key enumeration, when all parity nodes are down, which is rare case.
Hmm, it needs to merge the keys from all shards, if we use hash to compare, then it would need calculate hash for all keys, it is a lot overhead as well.

Reply:
OK. But here is to check if the same key already existed right? why above code check size match but key not match?
should use "&& strncmp(key_one->key.iov_buf, key, key_size) == 0" ?